### ISSUE TITLE
Support for other versions?

### ISSUE THREAD

**Issue by joelatgrayv** [miqwit/dedex#26]
Do you have plans for supporting other versions?

341, 381, 383, 430

Thanks for making this project. Also, I'm curious what was your motivation for making it?

**Comment by miqwit** [comment]
Hello @joelatgrayv. Thank you for your interest in dedex.

I actually have plans to support other versions, in particular later versions than the currently supported (3.8.2 and 4.1).

Note that the consortium announced the "sunset" plan of some versions of the standard, as visible here: https://kb.ddex.net/about-ddex-standards/what's-new/

> On 24th February 2023, DDEX announced that the following standards will be “sun-set” from 2025 onwards:
> All versions of the Electronic Release Notification Message Suite standard (ERN) that have the main version number of 3 and the version 4.0 of that standard; 

In practice, many still use versions 3.8.*. This sunset policy may encourage companies to upgrade their versions. 

> Also, I'm curious what was your motivation for making it?

As a software developer in the music industry, I had to code DDEX formatters or parsers multiple times. On my free time I decided to come up with a convenient library (in PHP) to make me and other developers save time during this process. It might seems "easy" to code an XML parser, and even fun; that's without considering the complexity and the possibilities of DDEX standards... It's harder than it looks.

Do you have any particular need? I encourage you to share to experience around the library/standard.

Have a good day.

**Comment by joelatgrayv** [comment]
Nice, thanks!

Would you be wiling to accept a PR for 341? Also, do you have any tips for adding that support for it?

**Comment by miqwit** [comment]
Sure, be my guest.

I suggest to start with one (or more) DDEX 341 examples that you add in the test suite; then write tests that parse this file and gets some data.

Don't commit private data in the XML file, like sender/recipients names and party IDs.

You'll also need to find the 341 XSD file to generate the PHP classes (I explain it in the official documentation).

Please read the README and CONTRIBUTE files and try to push a first version of your PR. We can then discuss in the PR.

**Comment by miqwit** [comment]
Hi @joelatgrayv . I just merged support for 43 (coded by edvard). And I did 42 right before.
As per older versions, like the ones you mention: 341, 381, 383; I am still interested to implement it. If you have sample files you can share, it will be helpful for the implementation (in particular the testing).

Do you have a bit of time to contribute to this (with me as a lead) adding and testing new versions?

**Comment by joelatgrayv** [comment]
@miqwit we have a fork that we have been using if you want to take a look and see if it’s worth merging. I think we hit some testing issues or something that prevented us from being able to get it approved. I can’t recall. 

https://github.com/SauceIndustries/dedex-fork

Otherwise, if you want to provide some guidance to @carlos-grayv and/or @GrayV31415265 with what you would like to see added, we would be happy to help. We can put it the schedule.

**Comment by miqwit** [comment]
Ah yes ok I see. I didn't realise you were from the same organization.
I know this PR, but I decided to not merge as it is, or iterate from it, juste because it's enormous.
I decided rather to ingest versions one by one, with corresponding unit tests.

Indeed, I asked for unit tests for each version, since I'd like to keep this repo as clean as I can ; and also because I may refactor some parts later one, and I really want the rest of the features to pass (regression testing).

I will likely cherry pick some of your code, or at least have a look, for previous versions of ERN and integrate them little by little in DeDEX. If I can count on some of you for some review, that will be very appreciated.

Also, I am amazed of how many versions you're handling. I often had to support 2 or 3 versions in my previous projects. Pretty cool!

**Comment by joelatgrayv** [comment]
Yeah of course, no problem. If you need anything from any of us, just ask. We would love to help.
